core/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.profile;
+
+import org.elasticsearch.search.profile.query.QueryProfileBreakdown;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingDeque;
+
+public abstract class AbstractInternalProfileTree<PB extends AbstractProfileBreakdown<?>, E> {
+
+    protected ArrayList<PB> timings;
+    /** Maps the Query to it's list of children.  This is basically the dependency tree */
+    protected ArrayList<ArrayList<Integer>> tree;
+    /** A list of the original queries, keyed by index position */
+    protected ArrayList<E> elements;
+    /** A list of top-level "roots".  Each root can have its own tree of profiles */
+    protected ArrayList<Integer> roots;
+    /** A temporary stack used to record where we are in the dependency tree.  Only used by scoring queries */
+    protected Deque<Integer> stack;
```

Comment:
The comment was not updated after the refactoring to make this class abstract. This stack is used to record the position in the dependency tree and is needed for both the query and agg impls